src/Adapter/MSTest.TestAdapter/Execution/TestExecutionManager.cs

```diff
@@ -485,12 +488,18 @@ private async Task ExecuteTestsInSourceAsync(IEnumerable<TestCase> tests, IRunCo
             // Queue the non parallel set
             if (nonParallelizableTestSet != null)
             {
-                await ExecuteTestsWithTestRunnerAsync(nonParallelizableTestSet, frameworkHandle, source, sourceLevelParameters, testRunner, usesAppDomains);
+                _testRunCancellationToken?.ThrowIfCancellationRequested();
+                // Ensure we can abandon currently running tasks on cancellation, rather than waiting for them to complete.
+                // They will still run on background but we won't await them.
+                await Task.WhenAny(ExecuteTestsWithTestRunnerAsync(nonParallelizableTestSet, frameworkHandle, source, sourceLevelParameters, testRunner, usesAppDomains), completeWhenTestRunIsCanceled);
```

Comment:
The more I think of it, the more I think it's probably not the right thing. Cancellation will be requested, then we just proceed while the task of `ExecuteTestsWithTestRunnerAsync` could still be running.

Reply:
I agree...you don't know what the dangling tasks will do and when...at this point it's better `Env.Exit(abort code)` the result of the run cannot be trusted and some other crash could happen for race for unexpected app state.